### ISSUE TITLE
npm update seems to collide with itself

### ISSUE THREAD

**Issue by aredridel** [npm/npm#9474]
```
:; npm update tap
npm ERR! gentlyRm containing path /Users/aredridel/Projects/npm/newww/node_modules/nyc/node_modules/rimraf isn't under npm's control
- minimatch@2.0.10 node_modules/nyc/node_modules/minimatch
- glob@5.0.14 node_modules/nyc/node_modules/glob
- minimist@0.0.8 node_modules/nyc/node_modules/minimist
- mkdirp@0.5.1 node_modules/nyc/node_modules/mkdirp
- nyc@2.4.0 node_modules/nyc
rimraf@2.4.3 node_modules/nyc/node_modules/rimraf -> node_modules/tap/node_modules/rimraf
newww@2.1.0 /Users/aredridel/Projects/npm/newww
├── gulp@3.9.0
├── gulp-nodemon@1.0.5
├── imagemin-pngcrush@1.0.0
├── numbat-emitter@1.1.3
├── numbat-process@2.0.7
└── request@2.60.0

npm ERR! Darwin 15.0.0
npm ERR! argv "node" "/usr/local/bin/npm" "update" "tap"
npm ERR! node v0.12.7
npm ERR! npm  v3.3.1
npm ERR! path /Users/aredridel/Projects/npm/newww/node_modules/nyc/node_modules/.bin/rimraf
npm ERR! code EEXIST

npm ERR! Refusing to delete: /Users/aredridel/Projects/npm/newww/node_modules/nyc/node_modules/.bin/rimraf not in /Users/aredridel/Projects/npm/newww/node_modules/nyc/node_modules/rimraf
File exists: /Users/aredridel/Projects/npm/newww/node_modules/nyc/node_modules/.bin/rimraf
Move it away, and try again.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/aredridel/Projects/npm/newww/npm-debug.log
```

Debug log:

https://gist.github.com/iarna/14edc840c236b7b872ec

**Comment by othiym23** [comment]
I would _really_ love to get the `install` and `update` code paths unified (which really means unifying `ls` and `outdated`, at least in part). Based on what @iarna has shown me of the code, I believe I know what's going on here, but I'm not sure. This too is an issue for her, and this one feels higher-priority to me than #9473. We'll decide during our next triage meeting whether it's a blocker or not. Thanks for filing these, @aredridel!

**Comment by othiym23** [comment]
@aredridel your good friend and mine @iarna is going to dig into this further. For now, I'm marking this as a blocker, and leaving it to Rebecca's discretion to remove the label once she's got a feeling for how severe an issue this is. How many times have you run into this? / How reproducible is this?

**Comment by iarna** [comment]
I thought I had an idea on how to make this happen. I _saw_ this happen once. But now, no, nope, no can reproduce, grr.

**Comment by iarna** [comment]
This may or may not be related to #9198. I mean, it's in the same code space.

**Comment by iarna** [comment]
Next up, to manufacture a tree that results in the same diff-set as the log... 😒

**Comment by ADTC** [comment]
My whole repo broke because of such complaints. Why can't `npm` delete the file? It's Windows btw, I hardly think permissions play a role here. What is the meaning of "file isn't under npm's control"?

It's always some file in a `.bin` folder, either a shell script or a `.bat` file.

**Comment by radek-holy** [comment]
I hit this always when deploying to Azure. My deployment script includes `npm prune --production` which fails with this error. The simplest reproducer is to create a new Azure App Service instance, log in using the Kudu's console, run `npm install mkdirp` and then `npm rm mkdirp`.

**Comment by PyDeq** [comment]
I hit this always when deploying to Azure. My deployment script includes `npm prune --production` which fails with this error. The simplest reproducer is to create a new Azure App Service instance, log in using the Kudu's console, run `npm install mkdirp` and then `npm rm mkdirp`.

**Comment by iarna** [comment]
This is/was an issue in `gently-rm`. We had additional fixes to it in v3.6.0` (from January 2016), so a further repro with a current version of npm would be much appreciated.
